packages/types/index.d.ts

```diff
@@ -1,84 +1,197 @@
+   *   `callback` is invoked with `END`
+   * - Otherwise `callback` is invoked with all buffered messages.
+   */
+  flush(cb: (items: T[] | END) => void): void
+  /**
+   * Closes the channel which means no more puts will be allowed. All pending
+   * takers will be invoked with `END`.
+   */
+  close(): void
 }
 
-export interface Effect<T, P> {
-  '@@redux-saga/IO': true;
-  combinator: false;
-  type: T;
-  payload: P;
-}
+export type Effect<T = any> = SimpleEffect<T, any> | CombinatorEffect<T, any>
 
-export type CombinedEffects<E> = {[key: string]: E} | E[];
+export type StrictEffect<T = any> = SimpleEffect<T, any> | StrictCombinatorEffect<T>
 
-export interface CombinatorEffect<T, E> {
-  '@@redux-saga/IO': true;
-  combinator: true;
-  type: T;
-  payload: CombinedEffects<E>;
+export interface StrictCombinatorEffect<T> extends CombinatorEffect<T, StrictEffect<T>> {}
+
+export interface SimpleEffect<T, P> {
+  '@@redux-saga/IO': true
+  combinator: false
+  type: T
+  payload: P
 }
 
-export type AnyEffect = Effect<any, any>;
-export interface AnyCombinatorEffect extends CombinatorEffect<any, ValidEffect> {}
+/**
+ * `all` / `race` effects
+ */
+export interface CombinatorEffect<T, E> {
+  '@@redux-saga/IO': true
+  combinator: true
+  type: T
+  payload: CombinatorEffectDescriptor<E>
+}
 
-export type ValidEffect = AnyEffect | AnyCombinatorEffect;
+export type CombinatorEffectDescriptor<E> = { [key: string]: E } | E[]
 
-export type END = { type: '@@redux-saga/CHANNEL_END' };
+export type END = { type: '@@redux-saga/CHANNEL_END' }
 
 /**
- * Annotate return type of generators with `SagaIterator` to get strict
- * type-checking of yielded effects.
+ * The Task interface specifies the result of running a Saga using `fork`,
+ * `middleware.run` or `runSaga`.
  */
-export type SagaIterator = IterableIterator<ValidEffect>;
-
 export interface Task {
-  isRunning(): boolean;
-  isCancelled(): boolean;
-  result<T = any>(): T | undefined;
-  error(): any | undefined;
-  toPromise<T = any>(): Promise<T>;
-  cancel(): void;
-  setContext<C extends object>(props: Partial<C>): void;
+  /**
+   * Returns true if the task hasn't yet returned or thrown an error
+   */
+  isRunning(): boolean
+  /**
+   * Returns true if the task has been cancelled
+   */
+  isCancelled(): boolean
+  /**
+   * Returns task return value. `undefined` if task is still running
+   */
+  result<T = any>(): T | undefined
+  /**
+   * Returns task thrown error. `undefined` if task is still running
+   */
+  error(): any | undefined
+  /**
+   * Returns a Promise which is either:
+   * - resolved with task's return value
+   * - rejected with task's thrown error
+   */
+  toPromise<T = any>(): Promise<T>
+  /**
+   * Cancels the task (If it is still running)
+   */
+  cancel(): void
+  setContext<C extends object>(props: Partial<C>): void
+}
+
+// https://github.com/Microsoft/TypeScript/issues/5453#issuecomment-419680547
+export namespace TupleHelpers {
```

Comment:
is it used somewhere?

Reply:
Also - shouldnt namespaces be avoided?

Reply:
it's used in effect.d.ts, in general I'd prefer not using namespaces, you have also pulled in all helpers from the linked issue, but you are using only 3 of them, so probably we should remove unused ones (even if they might come handy in the future) - or you could maybe publish those types as a package and we could depend on it here? :)

Reply:
If you're ok with adding a dependency, there's a package that has similar helpers: https://github.com/ksxnodemodules/typescript-tuple

Reply:
I have no problems with adding dependencies - especially non-runtime ones 😄 Outsource all the things 😉

Reply:
Done!